ocaml/typing/typedecl.ml

```diff
@@ -2211,26 +2194,24 @@ let abstract_type_decl ~injective layout params =
   generalize_decl decl;
   decl
 
-let approx_type_decl sdecl_list =
+let approx_type_decl env sdecl_list =
   let scope = Ctype.create_scope () in
   List.map
     (fun sdecl ->
        let id = Ident.create_scoped ~scope sdecl.ptype_name.txt in
+       let path = Path.Pident id in
        let injective = sdecl.ptype_kind <> Ptype_abstract in
        let layout =
          (* We set legacy_immediate to true because you were already allowed
             to write [@@immediate] on declarations. *)
          layout_of_attributes_default ~legacy_immediate:true
-           ~reason:(Type_declaration (Pident id))
+           ~context:(Type_declaration path)
            ~default:(Layout.value ~why:Default_type_layout)
            sdecl.ptype_attributes
        in
+       TyVarEnv.reset ();
```

Comment:
If you make the suggested change to `transl_type_param_layout`, it's no longer necessary to reset the `TyVarEnv` here.